runtime/astar/src/precompiles.rs

```diff
@@ -65,108 +65,72 @@ impl Contains<RuntimeCall> for WhitelistedCalls {
-            _ => None,
-        }
-    }
-
-    fn is_precompile(&self, address: H160, gas: u64) -> IsPrecompileResult {
-        let assets_precompile =
-            match Erc20AssetsPrecompileSet::<R>::new().is_precompile(address, gas) {
-                IsPrecompileResult::Answer { is_precompile, .. } => is_precompile,
-                _ => false,
-            };
-
-        IsPrecompileResult::Answer {
-            is_precompile: assets_precompile || Self::used_addresses().any(|x| x == address),
-            extra_cost: 0,
-        }
-    }
-}
-
-fn hash(a: u64) -> H160 {
-    H160::from_low_u64_be(a)
+#[precompile_utils::precompile_name_from_address]
+pub type AstarPrecompilesSetAt<R, C> = (
+    // Ethereum precompiles:
+    // We allow DELEGATECALL to stay compliant with Ethereum behavior.
+    PrecompileAt<AddressU64<1>, ECRecover, EthereumPrecompilesChecks>,
+    PrecompileAt<AddressU64<2>, Sha256, EthereumPrecompilesChecks>,
+    PrecompileAt<AddressU64<3>, Ripemd160, EthereumPrecompilesChecks>,
+    PrecompileAt<AddressU64<4>, Identity, EthereumPrecompilesChecks>,
+    PrecompileAt<AddressU64<5>, Modexp, EthereumPrecompilesChecks>,
+    PrecompileAt<AddressU64<6>, Bn128Add, EthereumPrecompilesChecks>,
+    PrecompileAt<AddressU64<7>, Bn128Mul, EthereumPrecompilesChecks>,
+    PrecompileAt<AddressU64<8>, Bn128Pairing, EthereumPrecompilesChecks>,
+    PrecompileAt<AddressU64<9>, Blake2F, EthereumPrecompilesChecks>,
+    // Non-Astar specific nor Ethereum precompiles :
+    PrecompileAt<AddressU64<1024>, Sha3FIPS256, (CallableByContract, CallableByPrecompile)>,
+    PrecompileAt<
+        AddressU64<1025>,
+        Dispatch<R, DispatchFilterValidate<RuntimeCall, WhitelistedCalls>>,
+        // Not callable from smart contract nor precompiles, only EOA accounts
+        // TODO: test this without the gensis hack for blacklisted
+        (),
+    >,
+    PrecompileAt<AddressU64<1026>, ECRecoverPublicKey, (CallableByContract, CallableByPrecompile)>,
+    PrecompileAt<AddressU64<1027>, Ed25519Verify, (CallableByContract, CallableByPrecompile)>,
+    // Astar specific precompiles:
+    PrecompileAt<
+        AddressU64<20481>,
+        DappsStakingWrapper<R>,
+        (CallableByContract, CallableByPrecompile),
+    >,
+    PrecompileAt<
+        AddressU64<20482>,
+        Sr25519Precompile<R>,
+        (CallableByContract, CallableByPrecompile),
+    >,
+    PrecompileAt<
+        AddressU64<20483>,
+        SubstrateEcdsaPrecompile<R>,
+        (CallableByContract, CallableByPrecompile),
+    >,
+    PrecompileAt<
+        AddressU64<20484>,
+        XcmPrecompile<R, C>,
+        (
+            SubcallWithMaxNesting<1>,
+            CallableByContract,
+            CallableByPrecompile,
+        ),
+    >,
+);
+
+pub type AstarPrecompiles<R, C> = PrecompileSetBuilder<
+    R,
+    (
+        // Skip precompiles if out of range.
+        PrecompilesInRangeInclusive<
+            // TODO: what is the range for precompiles sets 1 - ?
```

Comment:
You should just define the ranges we use, IMO. We can expand it as we add new precompile addresses.